### PR TITLE
notes taken during Mitch's course, Analysis-based program transformations

### PR DESCRIPTION
**Pull request by gasche**
Ping @bennn @mhyee : feel free to add your own notes.

### CODE REVIEW

File: analysis-based program transformations/talk-notes.txt [hunk 1]
```diff
@@ -0,0 +1,289 @@
+# HOPL: Analysis-based program transformations
+Mitch Wand
+February 21, 2017
+
+The thesis of the course is that if you take a look at the history of
+papers you can find the deltas between one paper and the next. As
+Matthias pointed out early on, you can't trust the people as they
```
Comment (tonyg) [inline analysis-based program transformations/talk-notes.txt]: "... can't trust the *papers* as they appear out of order..."

File: analysis-based program transformations/talk-notes.txt [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+appear out-of-order, and you can't trust the people as they remember
+badly. It's a really hard problem to figure out what people do and
+what they did it, that's what historians do. I want to reject the idea
```
Comment (tonyg) [inline analysis-based program transformations/talk-notes.txt]: "...and when they did it,..."

**Comment by bennn** [comment]
ok, I'll add mine around 11pm tonight (if you decide to merge before then, I'll make a PR with the additions)

File: analysis-based program transformations/talk-notes.txt [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    then l ∈ ρ̂(x) and ρ' ⊨ ρ̂
+ 
+- for closures
+
+  ((λx.e)ˡ, ρ) ⊨ (A, ρ̂) iff
+    (where A is a set of λ-labels)
+
+    ł ∈ A and ρ ⊨ ρ̂
+
+Soudness:
```
Comment (mhyee) [inline analysis-based program transformations/talk-notes.txt]: Typo: Soudness -> Sou**n**dness

File: analysis-based program transformations/talk-notes.txt [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+As an intermediate representation P', I chose the infinite unrolling
+of Fortran: unroll all the while-loops of the program, you get
+a rational tree, if you label them node trees the way Mitch did, you
+get back a finite representation, the program dependence graph
+(almost); we called it the "semantic program dependence graph", as it
+wasn't exactly the same.
+
+We circulated a manuscript to them in which we pointed out that their
+copmiler were all incorrect. In their compiler, if you had
```
Comment (mhyee) [inline analysis-based program transformations/talk-notes.txt]: Typo: copmiler -> compiler

File: analysis-based program transformations/talk-notes.txt [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+  y = t1
+  [...]
+  y = t2
+  [...]
+  use(y)
+
+they would always drop (t1) as it was useles if there was no use-of-y
```
Comment (mhyee) [inline analysis-based program transformations/talk-notes.txt]: Typo: useles -> useles**s**

File: analysis-based program transformations/talk-notes.txt [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+theoretical guy, should stay out of compilers as I polluting their
+field. It turned out that the semantic PDG was exactly the same as
+SSA. Of course, I stayed away from Fortran compilers.
+
+We, mathematicians, are a strange civilization that landed on the
+planet of compiler design and optimization, and we were sending
+messages into the future. If we don't change our message, they can't
+interact with it.
+
+tSometimes we in the field of mathematical compilers discover ideas
```
Comment (mhyee) [inline analysis-based program transformations/talk-notes.txt]: Typo: tSometimes -> Sometimes

**Comment by mhyee** [comment]
My notes are not as detailed, so I don't have anything to add (other than pointing out typos).

**Comment by bennn** [comment]
Committed some edits https://github.com/nuprl/hopl-s2017/pull/8/commits/66c91a5f698056afcdd3b34d903bb89ceda55683

**Comment by gasche** [comment]
Do I correctly understand that this is now good to merge? Anything still to be done?

**Comment by bennn** [comment]
merged
https://github.com/nuprl/hopl-s2017/commit/c901ceb165300bf5dce32135446cf586077746ae
https://github.com/nuprl/hopl-s2017/commit/03b922df86aa547f86c991545729c1dffca9a48a

**Pull request closed without merging**